### ISSUE TITLE
Assertion function/exception does not satisfy null safety

### ISSUE THREAD

**Issue by EliteMasterEric** [HaxeFoundation/haxe#11296]
**Minimal reproduction of this issue:** https://try.haxe.org/#Cb6cD3c8

**Expected behavior:** Since `Assert.isNotNull()` throws an exception, which is caught and handled by the try/catch block, `takesNotNull()` will never be reached if `input` is null, thus null safety should be satisfied. However, the compiler provides the following:

```
[ERROR] Test.hx:24: characters 19-24

 24 |      takesNotNull(input);
    |                   ^^^^^
    | Null safety: Cannot pass nullable value to not-nullable argument "input" of function "takesNotNull".
```

**Application/Use case:** The example above is from when I was implementing a set of unit tests for my application, and I annotated them with `@:nullSafety` with the expectation that `Assert.isNotNull` would be recognized as a null safety sentinel.

**Comment by kLabz** [comment]
Null safety doesn't analyze code going through multiple function calls to find that path leading to an exception :/

Not sure if that's feasible in your use case, but inlining both `fail` and `isNotNull` makes it work.
https://try.haxe.org/#eC031E44

Edit: alternatively, using something like `input = Assert.isNotNull(input);`, where `isNotNull` either returns a non null value or throws, would work too because compiler can see the call will affect `input`. https://try.haxe.org/#58948Db4

**Comment by EliteMasterEric** [comment]
> Not sure if that's feasible in your use case, but inlining both fail and isNotNull makes it work.

EDIT: I was able to create a macro to add `AInline` to `Assert.fail()`, and that appears to solve it.

**Comment by RblSb** [comment]
Yeah, `Assert.isNotNull(input);` will not work if it's not inlined, we can think about some meta to mark methods that will make arguments as non-nullable after execution, this is something to think about if https://github.com/HaxeFoundation/haxe/issues/7866 will be handled
